rfd/0107-abandon-active-access-request.md

```diff
@@ -0,0 +1,103 @@
+---
+authors: Jeff Pihach (jeff.pihach@goteleport.com)
+state: draft
+---
+
+# RFD 0107 - Abandon Active Access Requests
+
+## Required Approvers
+
+- Engineering: @zmb3
+
+## What
+
+Allow users who are using elevated permissions via an access request to abandon
+them ahead of their expiry once they are no longer of use.
+
+## Why
+
+By reducing the amount of time that a user has elevated permissions we also
+reduce the time and surface area to which an attack could occur in the event
+the users device becomes compromised.
+
+## Success Criteria
+
+- A user is able to abandon an active access request ahead of their expiry time
+  via the CLI or UI.
+- Creating a lock to prevent further use of the existing certificate.
+- An abandoned access request is unrecoverable and unusable.
+
+## Details
+
+### CLI
+
+A new `abandon` sub command to `tsh request` will be added. This command will
+call an GRPC method `AbandonAccessRequest` to mark the access request as
+`ABANDONED` and to set a lock on the access request.
+
+### UI
+
+When viewing your access request, in addition to the unassume option, there
+will also be a new button made available to abandon the active access request.
+The server will call the GRPC method `AbandonAccessRequest` to mark the access
+request as `ABANDONED` and to set a lock on the access request.
+
+### Core
+
+```protobuf
+// authservice.proto
+service AuthService {
+  ...
+  rpc AbandonAccessRequest(RequestID) returns (google.protobuf.Empty);
+  ...
+}
+```
+
+A new `RequestState` will be added to Access requests to indicate that the
+request has been intentionally abandoned. This will allow us to indicate in the
+UI that the request has been abandoned by the user.
+
+```protobuf
+// types.proto
+
+// RequestState represents the state of a request for escalated privilege.
+enum RequestState {
+  // NONE variant exists to allow RequestState to be explicitly omitted
+  // in certain circumstances (e.g. in an AccessRequestFilter).
+  NONE = 0;
+  // PENDING variant is the default for newly created requests.
+  PENDING = 1;
+  // APPROVED variant indicates that a request has been accepted by
+  // an administrating party.
+  APPROVED = 2;
+  // DENIED variant indicates that a request has been rejected by
+  // an administrating party.
+  DENIED = 3;
+  // Abandoned variant indicates that the request has been intentionally
+  // abandoned by the target of the access request.
+  ABANDONED = 4;
+}
+```
+
+### Why is a lock required
+
+Once a certificate is issued, without a lock, it's assumed to be valid until it
```

Comment:
Having the new call perform the same certificate invalidation as `drop` seems like a reasonable compromise. I'll update to remove references to, and requirement of, a lock.

Reply:
`drop` doesn't invalidate existing certificates, `drop` reissues new certificates to the user, and `tsh` happens to overwrite the existing ones. What would prevent a user to "abandon a request", making it seem like everything's fine, and then just swapping out the old certs in, to continue using the elevated privileges, without a lock?

And if we have to add a lock anyway, why do we need a new state for the access request?

Reply:
> What would prevent a user to "abandon a request", making it seem like everything's fine, and then just swapping out the old certs in, to continue using the elevated privileges, without a lock?

Nothing, and I think that's fine. For establishing strict security invariants, use a lock. Think of this more as a principle of least privilege feature rather than a strong security feature.

One of the original proposals was limit the number of times a single access request can be used to generate credentials. The feature proposed here is basically that, but instead of a fixed number of times, the access request author can state when the request can no longer be used for generating new credentials.